cmd/state-tray/internal/menu/projects.go

```diff
@@ -0,0 +1,83 @@
+package menu
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ActiveState/cli/cmd/state-tray/internal/open"
+	"github.com/ActiveState/cli/internal/graph"
+	"github.com/ActiveState/cli/internal/logging"
+	"github.com/getlantern/systray"
+)
+
+type LocalProjectsUpdater struct {
+	menuItem *systray.MenuItem
+	items    []*localProjectsMenuItem
+}
+
+type localProjectsMenuItem struct {
+	menuItem *systray.MenuItem
+	project  *graph.Project
+	done     <-chan struct{}
+	cancel   context.CancelFunc
+}
+
+func NewLocalProjectsUpdater(menuItem *systray.MenuItem) *LocalProjectsUpdater {
+	return &LocalProjectsUpdater{menuItem, []*localProjectsMenuItem{}}
+}
+
+func (u *LocalProjectsUpdater) Update(projects []*graph.Project) {
+	for _, item := range u.items {
+		item.remove()
+	}
+
+	u.items = make([]*localProjectsMenuItem, len(projects))
+	for i, project := range projects {
+		item := newLocalProjectMenuItem(project, u.menuItem)
+		u.items[i] = item
+		item.startEventLoop()
+	}
+}
+
+func newLocalProjectMenuItem(project *graph.Project, menuItem *systray.MenuItem) *localProjectsMenuItem {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	mProject := menuItem.AddSubMenuItem(project.Namespace, "")
```

Comment:
Nit: Might make your code a bit cleaner and better separate concerns if you add a `AddChild()` to `LocalProjectsUpdater` and pass `localProjectsMenuItem` to it. Effectively that would call `menuItem.AddSubMenuitem()` from `AddChild()` while keeping `localProjectsMenuItem` focused strictly on the menuitem itself.

So `LocalProjectsUpdater` is the thing that ties everything together.

Reply:
I like this idea however I think it would require `AddChild()` to modify the passed in `localProjectsMenuItem` as it needs the returned type from `menuItem.AddSubMenuItem()`.

Reply:
Why does it need to modify it? Sounds like it just needs to read a property.

Reply:
Would still like to see this addressed. The menuitem being passed this way is awkward.